### DISCUSSION TITLE
How to get followers of user in an ordered manner?

### DISCUSSION THREAD

**Discussion by facinick** [prisma/prisma#7486] [Q&A] [answered]
TL;DR: prisma query for 

(get followers of user with id = 1 sorted by id/name/etc)
```sql
    select id, name from users
        where users.id in (select senderId from friends where receiverId = 1)
        order by id "asc"
```

where `users` table has `user(name, id)` rows, and `friends` table has `friends(id, senderId, receiverId)` rows where sender and receiver ids are user ids from users table.

------------------------------------------------------------------

I have two basic tables, `users` and `friends`. `friends` have follow `senderId`, follow `receiverId` at them minimum and a composite `id` as well. Below is the prisma schema I've created:
```prisma
    model User {
      name              String  
      iFollowList      Friends[]    @relation("iFollowList")
      followsMeList    Friends[]    @relation("followsMeList")
      id    Int    @id @default(autoincrement())
    }
    
    model Friends {
      senderId      Int
      receiverId    Int
      sender       User    @relation(name: "followsMeList", fields: [senderId], references: [id])
      receiver     User    @relation(name: "iFollowList", fields: [receiverId], references: [id])
      @@id([senderId, receiverId])
    }
```

I want to run this simple query:

QUERY: **Get all the followers (basically an array of user) for user with `id = 1` in some sorted order on user's keys like id.**

and this is my prisma query for the same(here I'm trying to find followers of user with id = 1 let's say) : *(the query is for user, and i user the root.id to write a resolver for it's followers)*
```ts
    const _user = await prisma.friends.findMany({
        where: {
            receiverId: _root.id,
        },
        include: {
            sender: true
        },
    });
    
    // Then I manually do the mapping like >>
    // I DON't WANT TO DO THIS MAPPING, HOW DO I GET THIS RESULT DIRECTLY? IS IT POSSIBLE?
    const _followers = _user.map(queryResult => ({
        id: queryResult.sender.id,
        name: queryResult.sender.name,
    }));
    
    return _followers;
```
now this is working but,

----------------------------------------------------------------------------------------------------------------------
**PROBLEM 1:**
----------------------------------------------------------------------------------------------------------------------

You see, by doing "include", the result comes as a list of records from "friends" table with prisma joining the sender for me, so result is something like:

    [{
        senderId,
        receiverId,
        id,
        sender: {
            ...the user object
            for sender 1
        }
    },
    {
        senderId,
        receiverId,
        id,
        sender: {
            ...the user object
            for sender 2
        }
    }...]

and then I manually extract an array of users by using js's array.prototype.map function.

What I want is just an array of objects from Users table that are the followers. 

How do I perform this? Please let me know if you need anything else apart from what's mentioned.. it's been over 2 weeks and I'm not able to move past this.

----------------------------------------------------------------------------------------------------------------------
**PROBLEM 2:**
----------------------------------------------------------------------------------------------------------------------

lastly, I ALSO WANT TO ORDER THE FOLLOWERS BY LET'S SAY ID, for which I have no means but to manually sort the `_followers` array. How do I order inside `include`? (assuming there is no other way to get followers other than how I am getting as of now)

**Comment by ryands17** [comment, marked as answer]
@facinick 👋 

For **Problem 1**, Prisma only returns the data as per the models defined so you would have to map this in your application logic itself.

For **Problem 2**, you can use `orderBy` in `include` as follows:

```ts
include: { sender: { orderBy: [{ id: 'asc }] } }
```

**Reply by ryands17** [reply]
Ohh I see. In this case, the sender is just a single side of the relation so to order the main output by it, you would need to use:

```ts
const _user = await prisma.friends.findMany({
    where: {
      receiverId: 1,
    },
    include: {
      sender: true,
    },
    orderBy: [{ senderId: 'asc' }],
  })
```

You would also need to set this in your `schema.prisma`:

```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["orderByRelation"]
  binaryTargets   = ["native"]
}
```

**Reply by facinick** [reply]
Hey @ryands17 ,

I need to orderBy using user's name, id and createdAt fields. In the above code How will I order by name? sender is a user Object but how do I use it's various fields for ordering...?

**Reply by ryands17** [reply]
Like this:

```ts
const _user = await prisma.friends.findMany({
    where: {
      receiverId: 1,
    },
    include: {
      sender: true,
    },
    orderBy: [{ sender: { name: 'asc' } }],
  })
```

Using <kbd>ctrl+space</kbd> will give you autocomplete so you can see all the values that can be passed.

**Reply by facinick** [reply]
@ryands17  Thankyou so much! It's working now, I'm good to go :), could you direct me to a prisma doc where I can find more about preview features?

**Reply by ryands17** [reply]
You can find them [here](https://www.prisma.io/docs/concepts/components/preview-features) :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
